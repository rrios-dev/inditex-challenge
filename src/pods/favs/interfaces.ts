export interface FavHeroItem {
  id: number;
  name?: string;
  image?: string;
}

export interface FavsState {
  items: FavHeroItem[];
}
