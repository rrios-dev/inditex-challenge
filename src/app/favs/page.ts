export { default } from '@/scenes/hero-favs/hero-favs.scene';
