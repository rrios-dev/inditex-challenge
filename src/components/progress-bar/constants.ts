import { ProgressBarState } from "./interfaces";

export const PROGRESS_BAR_DEFAULT_STATE: ProgressBarState = {
  executors: new Set<string>(),
};
