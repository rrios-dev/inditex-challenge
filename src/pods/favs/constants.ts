import { FavsState } from "./interfaces";

export const DEFAULT_FAVS_STATE: FavsState = {
    ids: [],
}

export const FAVS_STORAGE_KEY = 'favs';