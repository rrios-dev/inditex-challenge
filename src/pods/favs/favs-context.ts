import { Dispatch, SetStateAction, createContext } from "react";

import { FavsState } from "./interfaces";


export interface FavsContext {
    favs: FavsState;
    setFavs: Dispatch<SetStateAction<FavsState>>;
}

const favsContext = createContext<FavsContext | null>(null);

export default favsContext;