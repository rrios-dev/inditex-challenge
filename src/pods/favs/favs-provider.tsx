"use client";

import { ReactNode } from "react";
import { useLocalStorage } from "usehooks-ts";

import { DEFAULT_FAVS_STATE, FAVS_STORAGE_KEY } from "./constants";
import favsContext from "./favs-context";
import { FavsState } from "./interfaces";

interface FavsProviderProps {
  children: ReactNode;
}

const FavsProvider = ({ children }: FavsProviderProps) => {
  const [favs, setFavs] = useLocalStorage<FavsState>(
    FAVS_STORAGE_KEY,
    DEFAULT_FAVS_STATE
  );

  return (
    <favsContext.Provider value={{ favs, setFavs }}>
      {children}
    </favsContext.Provider>
  );
};

export default FavsProvider;
