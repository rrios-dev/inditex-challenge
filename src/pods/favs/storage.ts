'use client';

import { DEFAULT_FAVS_STATE, FAVS_STORAGE_KEY } from "./constants";
import { FavsState } from "./interfaces";

const storage = window.localStorage;

const favsStorage = {
    get: () => {
        const data = storage.getItem(FAVS_STORAGE_KEY)
        return data ? JSON.parse(data) as FavsState : DEFAULT_FAVS_STATE;
    },
    set: (favs: FavsState) => storage.setItem(FAVS_STORAGE_KEY, JSON.stringify(favs)),
    clear: () => storage.setItem(FAVS_STORAGE_KEY, JSON.stringify(DEFAULT_FAVS_STATE))
} as const;

export default favsStorage;