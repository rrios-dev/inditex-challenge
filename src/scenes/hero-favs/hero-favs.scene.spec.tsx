import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { FAVS_STORAGE_KEY } from "@/pods/favs/constants";
import FavsProvider from "@/pods/favs/favs-provider";

import HeroFinder from "./hero-favs.scene";

vi.mock("next/navigation", async (importOriginal) => ({
  ...((await importOriginal()) as any),
  useSearchParams: () => ({ get: () => "" }),
  useRouter: () => ({ push: () => {} }),
}));

describe("Hero Favs Scene", () => {
  it("Mount Scene with two favs", () => {
    localStorage.setItem(
      FAVS_STORAGE_KEY,
      JSON.stringify({
        items: [
          {
            id: 1009144,
            image:
              "http://i.annihil.us/u/prod/marvel/i/mg/6/20/52602f21f29ec.jpg",
            name: "A.I.M.",
          },
          {
            id: 1011164,
            image:
              "http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg",
            name: "Alvin Maker",
          },
        ],
      })
    );

    const { asFragment } = render(
      <FavsProvider>
        <HeroFinder />
      </FavsProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Mount Scene without result", () => {
    localStorage.setItem(
      FAVS_STORAGE_KEY,
      JSON.stringify({
        items: [],
      })
    );

    const { asFragment } = render(
      <FavsProvider>
        <HeroFinder />
      </FavsProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
