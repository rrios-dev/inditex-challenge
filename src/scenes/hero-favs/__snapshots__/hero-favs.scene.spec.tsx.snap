// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hero Favs Scene > Mount Scene with two favs 1`] = `
<DocumentFragment>
  <main
    class="_hero-favs_06173e _flex_6eb20e _dir-column_6eb20e _gap-6_6eb20e"
  >
    <h1
      class="sr-only"
    >
      Hero Favs
    </h1>
    <div
      class="_flex_6eb20e _dir-column_6eb20e _gap-4_6eb20e"
    >
      <div
        class="_search-input-container_2e7b85"
      >
        <img
          alt="search item"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/icons/search.svg"
          style="color: transparent;"
          width="12"
        />
        <label
          class="sr-only"
          for=":r0:"
        >
          Search
        </label>
        <input
          class="_search-input_2e7b85"
          id=":r0:"
          placeholder="Search"
          type="text"
        />
      </div>
      <p
        class="_typography_142002 _variant-caption_142002 color-fg"
      >
        2 RESULTS
      </p>
    </div>
    <ul
      class="_hero-selector_1d6cb9"
    >
      <li>
        <a
          class="_link_dbb513 _hero-card_5aadb9"
          href="/hero/1009144-a.i.m."
        >
          <img
            alt="Thumbnail image of the hero A.I.M."
            class="_hero-card-image_5aadb9"
            loading="lazy"
            src="https://i.annihil.us/u/prod/marvel/i/mg/6/20/52602f21f29ec.jpg"
          />
          <hr />
          <div
            class="_bar_5aadb9"
          >
            <p
              class="_typography_142002 _variant-body2_142002 color-fg _name_5aadb9"
            >
              A.I.M.
            </p>
            <button
              class="_button-base_31eab0 _fav-button_5aadb9"
            >
              <img
                alt="Full heart"
                data-nimg="1"
                decoding="async"
                height="10"
                loading="lazy"
                src="/icons/heart-full-black.svg"
                style="color: transparent;"
                width="12"
              />
            </button>
          </div>
        </a>
      </li>
      <li>
        <a
          class="_link_dbb513 _hero-card_5aadb9"
          href="/hero/1011164-alvin-maker"
        >
          <img
            alt="Thumbnail image of the hero Alvin Maker"
            class="_hero-card-image_5aadb9"
            loading="lazy"
            src="https://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available.jpg"
          />
          <hr />
          <div
            class="_bar_5aadb9"
          >
            <p
              class="_typography_142002 _variant-body2_142002 color-fg _name_5aadb9"
            >
              Alvin Maker
            </p>
            <button
              class="_button-base_31eab0 _fav-button_5aadb9"
            >
              <img
                alt="Full heart"
                data-nimg="1"
                decoding="async"
                height="10"
                loading="lazy"
                src="/icons/heart-full-black.svg"
                style="color: transparent;"
                width="12"
              />
            </button>
          </div>
        </a>
      </li>
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
    </ul>
  </main>
</DocumentFragment>
`;

exports[`Hero Favs Scene > Mount Scene without result 1`] = `
<DocumentFragment>
  <main
    class="_hero-favs_06173e _flex_6eb20e _dir-column_6eb20e _gap-6_6eb20e"
  >
    <h1
      class="sr-only"
    >
      Hero Favs
    </h1>
    <div
      class="_flex_6eb20e _dir-column_6eb20e _gap-4_6eb20e"
    >
      <div
        class="_search-input-container_2e7b85"
      >
        <img
          alt="search item"
          data-nimg="1"
          decoding="async"
          height="12"
          loading="lazy"
          src="/icons/search.svg"
          style="color: transparent;"
          width="12"
        />
        <label
          class="sr-only"
          for=":r1:"
        >
          Search
        </label>
        <input
          class="_search-input_2e7b85"
          id=":r1:"
          placeholder="Search"
          type="text"
        />
      </div>
      <p
        class="_typography_142002 _variant-caption_142002 color-fg"
      >
        0 RESULT
      </p>
    </div>
    <ul
      class="_hero-selector_1d6cb9"
    >
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
      <li
        data-block="true"
      />
    </ul>
  </main>
</DocumentFragment>
`;
